src_python/org_agg.py

```diff
@@ -0,0 +1,238 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+import socket
+import pickle
+import select
+import time
+import numpy as np
+import pandas as pd
+import sys
+import logging
+from timeit import default_timer as timer
+import gc
+import math
+import os
+
+# CONSTANTS
+HEADER_LENGTH = 10
+
+logging.basicConfig(
+    level=logging.INFO, 
+    format='%(asctime)s : [%(levelname)s]  %(message)s',
+    handlers=[
+        logging.FileHandler('aggregator.log'),
+        logging.StreamHandler(sys.stdout)
+    ]
+)
+
+class Aggregator:
+
+    def __init__(self, model, rounds , graph_id, num_orgs , ip, port):
+
+        # Parameters
+        self.IP = ip
+        self.PORT = port
+        self.NUM_CLIENTS = num_orgs
+        self.ROUNDS = rounds
+
+        # Global model
+        self.model_weights = model
+
+        self.weights = []
+        self.partition_sizes = []
+        self.training_cycles = 0
+
+        self.stop_flag = False
+
+        # List of sockets for select.select()
+        self.sockets_list = []
+        self.clients = {}
+        self.client_ids = {}
+
+        # Craete Aggregator socket
+        self.aggregator_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self.aggregator_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        self.aggregator_socket.bind((self.IP, self.PORT))
+        self.aggregator_socket.listen(self.NUM_CLIENTS)
+
+        self.sockets_list.append(self.aggregator_socket)
+
+    def update_model(self,new_weights,num_examples):
+        self.partition_sizes.append(num_examples)
+        self.weights.append(num_examples * new_weights)
+
+        logging.info(str(self.partition_sizes))
+        logging.info(str(self.weights))
+
+        if len(self.weights) == self.NUM_CLIENTS:
+
+            avg_weight = sum(self.weights) / sum(self.partition_sizes)
+
+            self.weights = []
+            self.partition_sizes = []
+
+            self.model_weights = avg_weight
+
+            self.training_cycles += 1
+
+            for soc in self.sockets_list[1:]:
+                self.send_model(soc)
+            
+            logging.info("___________________________________________________ Training round %s done ______________________________________________________", self.training_cycles)
+        
+    def send_model(self, client_socket):
+
+        if self.ROUNDS == self.training_cycles:
+            self.stop_flag = True
+
+        weights = np.array(self.model_weights)
+        weights_path = "./weights/agg" + str(self.training_cycles) + ".npy"
```

Comment:
Where do we store these weights? on JasmineGraph server's source folder?

Reply:
Added into separate weights folder inside jaminegraph. Changes in 1875888487e22c8e17cbadd8acc558b28d98ebf0

Reply:
Changed file location to localstore 7018408aafed215ed25a1fa8237e25b9f4c978cf

Reply:
Done